<a href="https://colab.research.google.com/github/RiH-137/Bank-Customer-Churn-/blob/main/Ann_impl_lec90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip list

Package                          Version
-------------------------------- ---------------------
absl-py                          1.4.0
aiohttp                          3.9.3
aiosignal                        1.3.1
alabaster                        0.7.16
albumentations                   1.3.1
altair                           4.2.2
annotated-types                  0.6.0
anyio                            3.7.1
appdirs                          1.4.4
argon2-cffi                      23.1.0
argon2-cffi-bindings             21.2.0
array-record                     0.5.0
arviz                            0.15.1
astropy                          5.3.4
astunparse                       1.6.3
async-timeout                    4.0.3
atpublic                         4.0
attrs                            23.2.0
audioread                        3.0.1
autograd                         1.6.2
Babel                            2.14.0
backcall                         0.2.0
beautifulsoup4                   4.12.3
bi

**Bank Customer Churn**

In [2]:
import pandas as pd

In [3]:
df=pd.read_csv("https://raw.githubusercontent.com/sagnikghoshcr7/Bank-Customer-Churn-Prediction/master/Churn_Modelling.csv")

In [4]:
df.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [8]:
df.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [9]:
df.drop(columns=['RowNumber', 'CustomerId','Surname'], inplace=True)

In [10]:
df.head(3)

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1

In [13]:
df['Geography']

0        France
1         Spain
2        France
3        France
4         Spain
         ...   
9995     France
9996     France
9997     France
9998    Germany
9999     France
Name: Geography, Length: 10000, dtype: object

In [16]:
df['Geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

the above shows that there are total 5014 accounts from France.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  int64  
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(2), int64(7), object(2)
memory usage: 859.5+ KB


In [19]:
df['Gender'].value_counts()

Male      5457
Female    4543
Name: Gender, dtype: int64

Basically one-hot-encoding transform string data into integer data form.

Imagine a bank has customers from different places (Geography) and genders (Male/Female). One-hot encoding helps turn these categories into a format a computer can understand for building a model to predict churn (leaving the bank).

Before One-Hot Encoding:

Geography: The data might have text entries like "France", "Germany", "Spain".
Gender: It could be "Male" or "Female".
The computer struggles with these because it prefers numbers. One-hot encoding creates a new set of columns for each category.

After One-Hot Encoding:

New Columns: There will be separate columns for each geography (e.g., France, Germany, Spain) and gender (Male, Female).
Values: Each new column will have values of 0 or 1.
Interpretation:
A customer from France will have 1 in the "France" column and 0 in all other geography columns.
A male customer will have 1 in the "Male" column and 0 in the "Female" column.
This way, the computer can understand the customer's location and gender using these new columns with 0s and 1s, making it easier to analyze if these factors influence customer churn.

Benefit: One-hot encoding helps machine learning models treat categories like gender and geography effectively for tasks like churn prediction.

In [21]:
#the above all are categorical features. Therefore now we use ine-hot encoding.

#one-hot encoding
df=pd.get_dummies(df, columns=["Geography", "Gender"], drop_first=True)


#female=1; male=0
#france=(1,0,0); spain=(0,1,0); germany=(0,0,1)

In [22]:
df.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_Germany  \
0               1        101348.88       1                  0   
1               1        112542.58       0                  0   
2               0        113931.57       1                  0   
3               0         93826.63       0                  0   
4               1         79084.10       0                  0   

   Geography_Spain  Gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

X-->Independent variable

Y--> dependent variable ('Excite')

In [25]:
#for X
df.drop(['Exited'], axis=1)
X=df.drop(['Exited'], axis=1)

In [26]:
X.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88                  0                0   
1               1        112542.58                  0                1   
2               0        113931.57                  0                0   
3               0         93826.63                  0                0   
4               1         79084.10                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0

In [28]:
y=df['Exited']
y

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64

TensorFlow in a nutshell: It's a toolbox that helps design and train computer models by simplifying the process of feeding data, training, and making adjustments.

Keras is like a special set of tools within TensorFlow that makes designing these computer models (plays) much easier.

```




In [29]:
import tensorflow

In [30]:
from tensorflow import keras

In [31]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [34]:
len(X.columns)

11

Sequential groups a linear stack of layers into a tf.keras.Model.

Sequential provides training and inference features on this model.

Maintain layers working in a sequential manner and also maintain the forward and backward propogation.

In [36]:
model=Sequential()

#dence is responsible for adding layers.

model.add(Dense(11,activation="sigmoid",input_dim=11))#input layer
model.add(Dense(11,activation="sigmoid"))#hidden layer:- inside this hidden layer we can take any number of dence but with less than 11.
model.add(Dense(1,activation="sigmoid"))#output layer
#there are tota 11 columns in the dataset and hence 11 input therefore 11 numbr of neurons and we need a single output therefore 1 neuron for the output.

In [37]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 11)                132       
                                                                 
 dense_7 (Dense)             (None, 11)                132       
                                                                 
 dense_8 (Dense)             (None, 1)                 12        
                                                                 
Total params: 276 (1.08 KB)
Trainable params: 276 (1.08 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


total values= input_layer * hidden_layer + biases

              11*11+1

              132

In [38]:
#splliting the data into train and test

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)

Standard scaling is a technique used to preprocess data before feeding it into a machine learning model. Imagine you have a bunch of ingredients for a recipe, but some are measured in cups and others in grams. It's hard to follow the recipe if everything isn't using the same measurement system. Standard scaling does something similar for machine learning models.

In [39]:
#in Neural network we use gradient decent optimizer and GD uses standard scaler

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scale = scaler.fit_transform(X_train)
X_test_scale = scaler.transform(X_test)

In [40]:
# now we have to compile the model with the optimizer
#more follow tenserflow documnetation

''''
model.compile(optimizer=, loss=, metrics=)
'''
model.compile(optimizer="Adam", loss="binary_crossentropy", metrics=["accuracy"])       #since we have a classification problem and hence we use loss=binary_crossentropy
                                                                        #here accuracy is training accuracy not the predicted or final accuracy

In [46]:
model.fit(X_train,y_train,batch_size=50,epochs=10,verbose=1)

Epoch 1/10
160/160 [==============================] - 1s 3ms/step - loss: 0.6295 - accuracy: 0.7316
Epoch 2/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5180 - accuracy: 0.7960
Epoch 3/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5088 - accuracy: 0.7960
Epoch 4/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5060 - accuracy: 0.7960
Epoch 5/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5062 - accuracy: 0.7960
Epoch 6/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5049 - accuracy: 0.7960
Epoch 7/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5040 - accuracy: 0.7960
Epoch 8/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5024 - accuracy: 0.7960
Epoch 9/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5021 - accuracy: 0.7960
Epoch 10/10
160/160 [==============================] - 0s 3ms/step - loss: 0.5015 - accuracy: 0.7960

In [44]:
history=model.fit(X_train_scale,y_train,batch_size=50,epochs=10,verbose=1)

'''
In machine learning, an epoch refers to one complete pass through the entire training dataset.
 It's like giving your machine learning model one big run-through of all the training data it has available.

'''


'''
10 Epochs = 10 Training Runs: With epochs=10, the model will go through the training data set a total of 10 times,
 updating its internal parameters (like weights and biases) each time to improve its performance.

'''


'''
verbose=1 controls how much information you see during the training process of your model.

'''

'''

when training models with TensorFlow or similar libraries,
 batch_size=50 refers to a hyperparameter that determines the number of data points used in one update of the model's internal parameters during training.
 batch_size=50 --> 50dataset for each epoch.

'''

Epoch 1/10
160/160 [==============================] - 1s 4ms/step - loss: 0.3748 - accuracy: 0.8457
Epoch 2/10
160/160 [==============================] - 1s 4ms/step - loss: 0.3723 - accuracy: 0.8462
Epoch 3/10
160/160 [==============================] - 1s 3ms/step - loss: 0.3696 - accuracy: 0.8487
Epoch 4/10
160/160 [==============================] - 0s 3ms/step - loss: 0.3670 - accuracy: 0.8497
Epoch 5/10
160/160 [==============================] - 0s 3ms/step - loss: 0.3645 - accuracy: 0.8503
Epoch 6/10
160/160 [==============================] - 0s 2ms/step - loss: 0.3620 - accuracy: 0.8526
Epoch 7/10
160/160 [==============================] - 0s 3ms/step - loss: 0.3600 - accuracy: 0.8531
Epoch 8/10
160/160 [==============================] - 0s 2ms/step - loss: 0.3577 - accuracy: 0.8564
Epoch 9/10
160/160 [==============================] - 0s 3ms/step - loss: 0.3556 - accuracy: 0.8579
Epoch 10/10
160/160 [==============================] - 0s 3ms/step - loss: 0.3537 - accuracy: 0.8566

'\n10 Epochs = 10 Training Runs: With epochs=10, the model will go through the training data set a total of 10 times,\n updating its internal parameters (like weights and biases) each time to improve its performance.\n\n\n'

In [45]:
history

As we can see by using standard scalling, our accuracy increases from 0.79 to 0.85

NOTE:- Here is the accuracy is the training accuracy not the final or predicted accuracy.

In [48]:
y_pred=model.predict(X_test_scale)

63/63 [==============================] - 0s 2ms/step


In [49]:
y_pred

array([[0.16669089],
       [0.21672904],
       [0.15678053],
       ...,
       [0.21354057],
       [0.10834195],
       [0.15611589]], dtype=float32)

In [50]:
#now we have to convert y_pred args value into 0 and 1


y_pred=y_pred.argmax(axis=-1)

In [51]:
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [53]:
from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.7975